File: daily/2024/2463_Minimum_Total_Distance_Traveled.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    long long minimumTotalDistance(vector<int>& robot, vector<vector<int>>& factory) {
        sort(robot.begin(), robot.end());
        sort(factory.begin(), factory.end());
        while(!factory.empty() && factory[0][1] == 0) factory.erase(factory.begin());
        int r = robot.size();
        int f = factory.size();
        vector<vector<long long>> dp(f, vector<long long>(r, 1e12));
        dp[0][0] = abs(factory[0][0] - robot[0]);
        for (int l = 1; l < min(r, factory[0][1]); l++){
            dp[0][l] = dp[0][l-1] + abs(factory[0][0] - robot[l]);
        }
        for (int i = 1; i < f; i++){
            for (int j = 0; j < r; j++){
                dp[i][j] = dp[i-1][j];
                int limit = factory[i][1];
                long long acc = 0;
                for (int r = 0; r < limit && j - r >= 0; r++){
                    acc += abs(factory[i][0] - robot[j-r]);
                    if (j - r - 1 < 0){
                        dp[i][j] = min(dp[i][j], acc);
                    }
                    else{
                        if (dp[i-1][j-r-1] == 1e12) continue;
                        dp[i][j] = min(dp[i][j], dp[i-1][j-r-1] + acc);
                    }
                }
                if (dp[i][j] == 1e12) break;
            }
        }
        return dp[f-1][r-1];
    }
};

int main(){
    vector<int> robot = {0,4,6};
    vector<vector<int>> factory = {{2,2},{6,2}};
    // vector<int> robot = {1,-1};
    // vector<vector<int>> factory = {{-2,1},{2,1}};
    // vector<int> robot = {1,-1};
    // vector<vector<int>> factory = {{-2,1},{2,1},{3,0}};
    Solution S;
    cout << S.minimumTotalDistance(robot, factory) << endl;
    return 0;
}

File: daily_python/2463_Minimum_Total_Distance_Traveled.py
import os
import sys
from typing import List

class Solution:
    def minimumTotalDistance(self, robot: List[int], factory: List[List[int]]) -> int:
        robot.sort()
        factory.sort()
        factory = [f for f in factory if f[1] > 0]
        r = len(robot)
        f = len(factory)
        prev = [float('inf')] * r
        curr = [float('inf')] * r
        prev[0] = abs(robot[0] - factory[0][0])
        for j in range(1, min(r,factory[0][1])):
            prev[j] = prev[j-1] + abs(robot[j] - factory[0][0])
        for i in range(1, f):
            for j in range(r):
                curr[j] = prev[j]
                limit = factory[i][1]
                acc = 0
                for k in range(limit):
                    if j - k >= 0:
                        acc += abs(factory[i][0] - robot[j-k])
                        if j - k - 1 < 0:
                            curr[j] = min(curr[j], acc)
                        else:
                            if prev[j-k-1] == float('inf'):
                                continue
                            curr[j] = min(curr[j], acc + prev[j-k-1])
                if curr[j] == float('inf'):
                    break
            prev = curr.copy()
        return prev[-1]
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            robot = list(map(int, lines[0].split('=')[1].strip()[1:-2].split(',')))
            factory = [list(map(int, i.split(','))) for i in lines[1].split('=')[1].strip()[2:-3].split('},{')]
            dataset.append((robot, factory))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for robot, factory in dataset:
        results.append(solution.minimumTotalDistance(robot, factory))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')